<html>
<head>
<link rel="stylesheet"  type="text/css" href="help.css">
</head>
<body>

<!-- HEADER -->

<h1>Error Correction</h1>
<p>
Sequences of class labels as obtained by the event selection can be corrected for errors by incorporation of time information.
<br/>
<br/>
Correction of errors is performed by a <img class="extlink" src=":extlink"/><a href="http://en.wikipedia.org/wiki/Hidden_Markov_model">hidden Markov model</a> (HMM), which is learned from all event-sequences of one experimental condition.
<br/>
The HMM is implemented in scripts of the <img class="extlink" src=":extlink"/><a href="http://www.r-project.org]">R-project</a> and requires installation of additional packages (this is going to change soon). CecogAnalyzer calls the R script, which reads the event-based trajectory files and generates the HMMs. The visualized results are written to one HTML page per channel, which opens automatically after processing is completed. The user can constrain graphs by specifying one text file per channel. The fully connected graph implying no constrains is used per default. The association between positions (individual movies) and experimental information such as oligo identifier or the name of the targeted gene symbol can also be defined in a simple text file (see example files in the generated <tt>hmm</tt> folder). Based on this information trajectories are grouped by position, oligo ID or gene symbol and HMMs are generated per group.<br>
Error correction can be tested separately for already picked events, e.g. by motif selection in <a href="qrc:/tracking">Tracking</a>.
</p>

<hr></hr>

<h3>Parameters</h3>

<a name="filename_to_r">
<h4>R-project executable</h4>
<p class="text">Filename of the R executable. If the R-project is not part of the system's <tt>PATH</tt> variable the full filename is required, e.g. <tt>C:\Program Files\R\R-2.10.0\bin\R.exe</tt> on Windows for R 2.10.0.
</p>
</a>

<a name="constrain_graph">
<h4>Constrain graph</h4>
<p class="text">A priori information about a cellular process can be applied by constraining the transition between cellular states. These transitions can be expressed as a graph structure and per default the fully connected graph (all connections between states are possible) is used. The transition probabilities of the HMM are learned according to the underlying graph (the probability is zero for all missing edges).
<br/>If this option is omitted the fully connected graph is used and written as text file to the <tt>hmm</tt> folder (which can be modified and used as example).


<a name="primary_graph">
<h4 class="sub">Primary file</h4>
<p class="subtext">Text file defining the graph structure for the classifier of the primary channel./p>
</a>

<a name="secondary_graph">
<h4 class="sub">Secondary file</h4>
<p class="subtext">Text file defining the graph structure for the classifier of the secondary channel.</p>
</a>
</p>
</a>

<a name="position_labels">
<h4>Position labels</h4>
<p class="text">Association of the experimental condition to a position (also named <i>plate mapping</i> or <i>plate layout</i>). For now only <i>oligo ID</i> and <i>gene symbol</i> are supported. If omitted the position names are used for identification of the experiment and a sample file <tt>&lt;OUTPUT&gt;/hmm/layout.txt</tt> is created.

<a name="mappingfile">
<h4 class="sub">File</h4>
<p class="subtext">Simple tab-separated text file with three columns: <tt>Position</tt> (must match the position names), <tt>OligoID</tt> (any string), and <tt>GeneSymbol</tt> (any string).</p>
</a>
</p>
</a>

<a name="groupby">
<h4>Group by</h4>
<p class="text">The HMM is learned and errors are corrected per experimental condition. How positions are grouped can be defined here based on the position labels above.

<a name="groupby_position">
<h4 class="sub">Position</h4>
<p class="subtext">Group by Position (no position labels necessary).</p>
</a>
<a name="groupby_oligoid">
<h4 class="sub">Oligo ID</h4>
<p class="subtext">Group by oligo ID (defined in the position labels).</p>
</a>
<a name="groupby_genesymbol">
<h4 class="sub">Gene symbol</h4>
<p class="subtext">Group by gene symbol (defined in the position labels).</p>
</a>
</p>
</a>

<a name="plot_parameter">
<h4>Plot parameter</h4>
<p class="text">Parameters for plot generation.

<a name="timelapse">
<h4 class="sub">Time-lapse [min]</h4>
<p class="subtext">Specifies the time-lapse.</p>
</a>
<a name="max_time">
<h4 class="sub">Max. time in plot [min]</h4>
<p class="subtext">The maximal time (y-axis) in a bar plot or box plot.</p>
</a>
</p>
</a>

<hr></hr>
<a name="controlpanel">
<h3>Control Panel</h3>
<p>
<ul>
<li><b>Correct errors</b> - Performs HMM error correction on event trajectories already present in the output sub-folders (e.g. generated either by <a href="qrc:/processing">Processing</a> or <a href="qrc:/tracking#controlpanel">Apply event selection</a> in Tracking. Error correction is only performed for primary and/or secondary channel if the corresponding classification is enabled in <a href="qrc:/processing">Processing</a>.</li>
</ul>
</p>
</a>


<!-- FOOTER -->

</body>
</html>